src/correction.cc

```diff
@@ -45,25 +96,186 @@ Formula::Formula(const rapidjson::Value& json) :
+    else if ( name == "erf" )   { fun = [](double x) { return std::erf(x); }; }
+    else if ( name == "sqrt" )  { fun = [](double x) { return std::sqrt(x); }; }
+    else if ( name == "abs" )   { fun = [](double x) { return std::abs(x); }; }
+    else if ( name == "cos" )   { fun = [](double x) { return std::cos(x); }; }
+    else if ( name == "sin" )   { fun = [](double x) { return std::sin(x); }; }
+    else if ( name == "tan" )   { fun = [](double x) { return std::tan(x); }; }
+    else if ( name == "acos" )  { fun = [](double x) { return std::acos(x); }; }
+    else if ( name == "asin" )  { fun = [](double x) { return std::asin(x); }; }
+    else if ( name == "atan" )  { fun = [](double x) { return std::atan(x); }; }
+    else if ( name == "cosh" )  { fun = [](double x) { return std::cosh(x); }; }
+    else if ( name == "sinh" )  { fun = [](double x) { return std::sinh(x); }; }
+    else if ( name == "tanh" )  { fun = [](double x) { return std::tanh(x); }; }
+    else if ( name == "acosh" ) { fun = [](double x) { return std::acosh(x); }; }
+    else if ( name == "asinh" ) { fun = [](double x) { return std::asinh(x); }; }
+    else if ( name == "atanh" ) { fun = [](double x) { return std::atanh(x); }; }
+    else {
+      throw std::runtime_error("unrecognized unary function: " + std::string(name));
+    }
+    return {
+      Ast::NodeType::UnaryCall,
+      fun,
+      {translate_ast(*ast.nodes[1])}
+    };
+  }
+  else if (ast.name == "CALLB" ) {
+    if ( ast.nodes.size() != 3 ) { throw std::runtime_error("CALLB without 3 nodes?"); }
+    Ast::BinaryFcn fun;
+    auto name = ast.nodes[0]->token;
+    if      ( name == "atan2" ) { fun = [](double x, double y) { return std::atan2(x, y); }; }
+    else if ( name == "pow" )   { fun = [](double x, double y) { return std::pow(x, y); }; }
+    else if ( name == "max" )   { fun = [](double x, double y) { return std::max(x, y); }; }
+    else if ( name == "min" )   { fun = [](double x, double y) { return std::min(x, y); }; }
+    else {
+      throw std::runtime_error("unrecognized binary function: " + std::string(name));
+    }
+    return {
+      Ast::NodeType::BinaryCall,
+      fun,
+      {translate_ast(*ast.nodes[1]), translate_ast(*ast.nodes[2])}
+    };
+  }
+  else if (ast.name == "EXPRESSION" ) {
+    if ( ast.nodes.size() != 3 ) { throw std::runtime_error("EXPRESSION without 3 nodes?"); }
+    return {
+      Ast::NodeType::Expression,
+      ast.nodes[1]->token[0],
+      {translate_ast(*ast.nodes[0]), translate_ast(*ast.nodes[2])}
+    };
+  }
+  throw std::runtime_error("Unrecognized AST node");
+}
+
+double Formula::evaluate(const std::vector<Variable>& inputs, const std::vector<Variable::Type>& values) const {
+  std::vector<double> variables;
+  variables.reserve(variableIdx_.size());
+  for ( auto idx : variableIdx_ ) { variables.push_back(std::get<double>(values[idx])); }
+  if ( ! ast_ ) build_ast();
```

Comment:
This makes Formula::evaluate non-reentrant and therefore not thread safe.

Reply:
Shoot. Digging for idioms, I guess what I want is a thread safe lazy initialization. Seem `std::call_once` fits the bill?

Reply:
The absolute best is to just do the ast building in the constructor. What is it about the usage pattern which is driving you not to do that? In general, delayed initialization in a threaded context is a very bad thing to do.

Reply:
There are sets of B-tagging scale factors that are a binned lookup into 18k (!!) formulas. Usually the user accesses only a portion of the phase space, so not parsing every formula can save time. Dramatically so for TFormula: using https://github.com/cms-sw/cmssw/blob/master/CondTools/BTau/interface/BTagCalibrationReader.h on the csv equivalent to the example file in this repo takes 5 minutes just to construct. That said, this PEG parser is really fast and can compile all 18k in about 2s so the motivation behind delayed init is reduced. Further, these formulas follow a regular pattern and I think if BTV adopted the JME solution of replacing float literals with parameters then there would only be a handful of formulas to compile.